flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/OperatorCoordinatorHolder.java

```diff
@@ -299,9 +374,73 @@ public void resetToCheckpoint(long checkpointId, @Nullable byte[] checkpointData
+            long checkpointId, CompletableFuture<byte[]> result) {
         mainThreadExecutor.assertRunningInMainThread();
 
+        try {
+            if (currentPendingCheckpointId != OperatorCoordinator.NO_CHECKPOINT
+                    && currentPendingCheckpointId != checkpointId) {
+                throw new IllegalStateException(
+                        String.format(
+                                "Cannot checkpoint coordinator for checkpoint %d, "
+                                        + "since checkpoint %d has already started.",
+                                checkpointId, currentPendingCheckpointId));
+            }
+
+            if (latestAttemptedCheckpointId >= checkpointId) {
+                throw new IllegalStateException(
+                        String.format(
+                                "Regressing checkpoint IDs. Previous checkpointId = %d, new checkpointId = %d",
+                                latestAttemptedCheckpointId, checkpointId));
+            }
+
+            subtaskGatewayMap.forEach(
+                    (subtask, gateway) -> gateway.markForCheckpoint(checkpointId));
+
+            Preconditions.checkState(acknowledgeCloseGatewayFutureMap.isEmpty());
+        } catch (Throwable t) {
+            result.completeExceptionally(t);
+            globalFailureHandler.handleGlobalFailure(t);
+            return;
+        }
+
+        currentPendingCheckpointId = checkpointId;
+        latestAttemptedCheckpointId = checkpointId;
+
+        for (int subtask : subtaskGatewayMap.keySet()) {
+            acknowledgeCloseGatewayFutureMap.put(subtask, new CompletableFuture<>());
+            final OperatorEvent closeGatewayEvent = new CloseGatewayEvent(checkpointId, subtask);
+            subtaskGatewayMap
+                    .get(subtask)
+                    .sendEventWithCallBackOnCompletion(
+                            closeGatewayEvent,
+                            (success, failure) -> {
+                                if (failure != null) {
+                                    // If the close gateway event failed to reach the subtask for
+                                    // some reason, the coordinator would trigger a fail-over on
+                                    // the subtask if the subtask is still running. If the subtask
+                                    // has been not running, it is guaranteed that the coordinator
+                                    // won't receive more events from this subtask before the
+                                    // current checkpoint finishes, which is equivalent to receiving
+                                    // ACK from this subtask.
+                                    if (!(failure instanceof TaskNotRunningException)) {
+                                        subtaskGatewayMap
+                                                .get(subtask)
+                                                .triggerTaskFailover(closeGatewayEvent, failure);
```

Comment:
Would it be simpler to move this logic to `SubtaskGatewayImpl::sendEventWithCallBackOnCompletion(...)`?